examples/vision/forwardforward.py

```diff
@@ -0,0 +1,352 @@
+                goodness += [tf.math.reduce_mean(tf.math.pow(h, 2), 1)]
+            goodness_per_label += [
+                tf.expand_dims(tf.reduce_sum(goodness, keepdims=True), 1)
+            ]
+        goodness_per_label = tf.concat(goodness_per_label, 1)
+        return tf.argmax(goodness_per_label, 1)
+
+    def predict(self, data):
+        x = data
+        preds = list()
+        for idx in trange(x.shape[0]):
+            sample = x[idx]
+            result = self.predict_one_sample(sample)
+            preds.append(result)
+        return np.asarray(preds, dtype=int)
+
+    def train_step(self, data):
+        x, y = data
+        x = self.flatten_layer(x)
+        perm_array = tf.range(start=0, limit=x.get_shape()[0], delta=1)
+        x_pos = self.overlay_y_on_x(x, y)
+        y_numpy = y.numpy()
+        random_y_tensor = y_numpy[tf.random.shuffle(perm_array)]
+        x_neg = self.overlay_y_on_x(x, tf.convert_to_tensor(random_y_tensor))
+        h_pos, h_neg = x_pos, x_neg
+        for idx, layer in enumerate(self.layers):
+            if idx == 0:
+                print("Input layer : No training")
+                continue
+            print(f"Training layer {idx+1} now : ")
+            if isinstance(layer, FFDense):
+                h_pos, h_neg, loss, loss_list = layer.forwardforward(h_pos, h_neg)
+                plt.plot(range(len(loss_list)), loss_list)
+                plt.title(f"Loss over training on layer {idx+1}")
+                plt.show()
+            else:
+                x = layer(x)
+        return {"FinalLoss": loss}
+
+
+"""
+## Convert MNIST `NumPy` arrays to `tf.data.Dataset`
+
+We now perform some preliminary processing on the `NumPy` arrays and then convert them
+into the `tf.data.Dataset` format which allows for optimized loading.
+"""
+
+x_train = x_train.astype(float) / 255
+x_test = x_test.astype(float) / 255
+
+train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+test_dataset = tf.data.Dataset.from_tensor_slices((x_test, y_test))
+
+train_dataset = train_dataset.batch(60000)
+test_dataset = test_dataset.batch(10000)
+
+"""
+## Fit the network and visualize results
+
+Having performed all previous set-up, we are now going to run `model.fit()` and run 1
+model epoch, which will perform 2000 epochs on each layer. We get to see the plotted loss
+curve as each layer is trained.
+"""
+
+model = FFNetwork(dims=[784, 500, 500])
+
+model.compile(
+    optimizer=keras.optimizers.Adam(learning_rate=0.03), loss="mse", run_eagerly=True
+)
+
+history = model.fit(train_dataset, epochs=1)
+
+"""
+## Perform inference and testing
+
+Having trained the model to a large extent, we now see how it performs on the test set.
+We calculate the Accuracy Score to understand the results closely.
+"""
+
+preds = model.predict(tf.convert_to_tensor(x_test))
+
+preds = preds.reshape((preds.shape[0], preds.shape[1]))
+
+results = accuracy_score(preds, y_test)
+
+print(f"Accuracy score : {results*100}%")
```

Comment:
What do you get?

Reply:
The Accuracy score currently stood at ~24-27% on different runs (can set a constant Seed to enhance reproducibility)

Reply:
25% accuracy on MNIST tells you that your algorithm does not work, unfortunately. A simple logistic regression does ~92% or so. A simple logistic regression after applying ~90% noise on the input (setting to 0 90% of the pixels, randomly) still does 67%. Even if it were poorly tuned, the example should achieve at least 97% if the algorithm did work as expected.

Reply:
Right. Let me iterate on this algorithm for a while, see if I am making some errors and I will get back to you on this